clients/pkg/promtail/targets/objectstore/aws.go

```diff
@@ -0,0 +1,198 @@
+	Size int64
+}
+
+var (
+	defaultMaxRetries          = 5
+	defaultMaxNumberOfMessages = int64(10)
+	defaultVisibilityTimeout   = int64(20)
+)
+
+func newS3Client(cfg SQSConfig) (Client, error) {
+	sess, err := buildSQSClient(cfg)
+	if err != nil {
+		return nil, err
+	}
+	svc := sqs.New(sess)
+	urlResult, err := svc.GetQueueUrl(&sqs.GetQueueUrlInput{
+		QueueName: &cfg.Queue,
+	})
+	if err != nil {
+		return nil, err
+	}
+	queueURL := urlResult.QueueUrl
+	return &s3Client{
+		queueURL: queueURL,
+		svc:      svc,
+	}, nil
+}
+
+func buildSQSClient(cfg SQSConfig) (*session.Session, error) {
+	var sqsConfig *aws.Config
+
+	sqsConfig = sqsConfig.WithMaxRetries(defaultMaxRetries)
+
+	if cfg.Region != "" {
+		sqsConfig = sqsConfig.WithRegion(cfg.Region)
+	}
+
+	if cfg.AccessKeyID != "" && cfg.SecretAccessKey == "" ||
+		cfg.AccessKeyID == "" && cfg.SecretAccessKey != "" {
+		return nil, errors.New("must supply both an Access Key ID and Secret Access Key or neither")
+	}
+
+	if cfg.AccessKeyID != "" && cfg.SecretAccessKey != "" {
+		creds := credentials.NewStaticCredentials(cfg.AccessKeyID, cfg.SecretAccessKey, "")
+		sqsConfig = sqsConfig.WithCredentials(creds)
+	}
+
+	tlsConfig := &tls.Config{
+		InsecureSkipVerify: cfg.HTTPConfig.InsecureSkipVerify,
+	}
+
+	if cfg.HTTPConfig.CAFile != "" {
+		tlsConfig.RootCAs = x509.NewCertPool()
+		data, err := os.ReadFile(cfg.HTTPConfig.CAFile)
+		if err != nil {
+			return nil, err
+		}
+		tlsConfig.RootCAs.AppendCertsFromPEM(data)
+	}
+
+	transport := http.RoundTripper(&http.Transport{
+		Proxy: http.ProxyFromEnvironment,
+		DialContext: (&net.Dialer{
+			Timeout:   30 * time.Second,
+			KeepAlive: 30 * time.Second,
+			DualStack: true,
+		}).DialContext,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       cfg.HTTPConfig.IdleConnTimeout,
+		MaxIdleConnsPerHost:   100,
+		TLSHandshakeTimeout:   3 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
+		ResponseHeaderTimeout: cfg.HTTPConfig.ResponseHeaderTimeout,
+		TLSClientConfig:       tlsConfig,
+	})
+
+	httpClient := &http.Client{
+		Transport: transport,
+	}
+
+	sqsConfig = sqsConfig.WithHTTPClient(httpClient)
+
+	sess, err := session.NewSession(sqsConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create new s3 session")
+	}
+	return sess, nil
+}
+
+func (s3Client *s3Client) ReceiveMessage(timeout int64) ([]messageObject, error) {
+	msgResult, err := s3Client.svc.ReceiveMessage(&sqs.ReceiveMessageInput{
+		AttributeNames: []*string{
+			aws.String(sqs.MessageSystemAttributeNameSentTimestamp),
+		},
+		MessageAttributeNames: []*string{
+			aws.String(sqs.QueueAttributeNameAll),
+		},
+		QueueUrl:            s3Client.queueURL,
+		MaxNumberOfMessages: aws.Int64(defaultMaxNumberOfMessages), // hard coded. We might take this as user input in future versions.
+		WaitTimeSeconds:     aws.Int64(timeout),
+		VisibilityTimeout:   aws.Int64(defaultVisibilityTimeout), // hard coded. We might take this as user input in future versions.
```

Comment:
If a large file takes more than 20sec to process and ack does it means another message will be sent. This means you have to run a single Promtail right ? if you run more then the position file is not shared/sync and they will both comsume the same message ?

Reply:
Yeah, to scale to process more files, running Promtail with same SQS queue would be tricky. One solution would be to create multiple SQS queues based on some pattern like `bucket/a/file.gz` will go to queue A and `bucket/b/file.gz` will go to queue B. Then multiple Promtail can be configured with different queues (A, B etc...)